kotlinx-coroutines-core/common/src/flow/internal/Combine.kt

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+@file:Suppress("UNCHECKED_CAST", "NON_APPLICABLE_CALL_FOR_BUILDER_INFERENCE") // KT-32203
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.internal.Symbol
+import kotlinx.coroutines.selects.*
+
+internal fun getNull(): Symbol = NULL // Workaround for JS BE bug
+
+internal suspend inline fun <T1, T2, R> FlowCollector<R>.combineTransformInternal(
+    first: Flow<T1>, second: Flow<T2>,
+    crossinline transform: suspend FlowCollector<R>.(a: T1, b: T2) -> Unit
+) {
+    coroutineScope {
+        val firstChannel = asFairChannel(first)
+        val secondChannel = asFairChannel(second)
+        var firstValue: Any? = null
+        var secondValue: Any? = null
+        var firstIsClosed = false
+        var secondIsClosed = false
+        while (!firstIsClosed || !secondIsClosed) {
+            select<Unit> {
+                onReceive(firstIsClosed, firstChannel, { firstIsClosed = true }) { value ->
+                    firstValue = value
+                    if (secondValue !== null) {
+                        transform(getNull().unbox(firstValue), getNull().unbox(secondValue) as T2)
+                    }
+                }
+
+                onReceive(secondIsClosed, secondChannel, { secondIsClosed = true }) { value ->
+                    secondValue = value
+                    if (firstValue !== null) {
+                        transform(getNull().unbox(firstValue) as T1, getNull().unbox(secondValue) as T2)
+                    }
+                }
+            }
+        }
+    }
+}
+
+@PublishedApi
+internal fun <T, R> combine(
+    vararg flows: Flow<T>,
+    arrayFactory: () -> Array<T?>,
+    transform: suspend FlowCollector<R>.(Array<T>) -> Unit
+): Flow<R> = flow {
```

Comment:
TooManyFunctionsCalledCombineException. I'd suggest to rename this one to `combineImpl`.

Reply:
`combineInternal` to be consistent with the rest of the API